Initial implementation for lyrics

kgarner7 opened the pull request:
This is a draft of a potential implementation for displaying lyrics (both synced and not). Currently only Navidrome, because it looks like the Jellyfin fetcher doesn't currently get lyrics?
Inspired by https://github.com/navidrome/react-music-player. It works best if you just start a song from zero and continue (and performs better in the electron app for some reason...); seeking has mixed results

kgarner7 committed:
initial implementation for lyrics

jeffvli commented:
To do:
 - [x] Add Jellyfin lyrics (referenced in #52)
   - [x] Support synced format
   - [x] Support unsynced format
 - [x] Add configurable fetchers for external lyrics (Reference fetchers from [Sunamu](https://github.com/NyaomiDEV/Sunamu/tree/master/src/main/lyricproviders))
   - [x] Netease
   - [x] Genius
 - [x] Add fancier animations for synced lyrics display (use framer motion)
 - [x] Add configurable offset for synced lyrics timing
 - [x] ...more tbd

kgarner7 committed:
Improved lyric syncing, fetch

- uses a somewhat more sane way to parse lyrics and teardown timeouts
- adds 'seeked' to setCurrentTime to make detecting seeks in lyric much easier
- adds ability to fetch lyrics from genius/netease (desktop only)

kgarner7 committed:
add jellyfin, improvements

kgarner7 commented:
All of the above have been added save for framer-motion (and tbd)

jeffvli commented:
Thanks, I'll review this over the weekend!

kgarner7 committed:
support .txt

jeffvli commented on src/renderer/features/settings/components/playback/lyric-settings.tsx:
```diff
@@ -0,0 +1,110 @@
+import { NumberInput, Switch } from '@mantine/core';
```
These need to be imported from `/@/renderer/components` rather than mantine.

jeffvli commented on src/renderer/features/lyrics/lyrics.tsx:
```diff
@@ -0,0 +1,131 @@
+const timeExp = /\[(\d{2,}):(\d{2})(?:\.(\d{2,3}))?]([^\n]+)\n/g;
+
+export const Lyrics = () => {
+  const currentServer = useCurrentServer();
```
The queue is independent of the current server, meaning you'll need to use `serverId` from the `currentSong` to determine what the actual server is.

jeffvli commented on src/renderer/features/lyrics/lyrics.tsx:
```diff
@@ -0,0 +1,131 @@
+  }, []);
+
+  useEffect(() => {
+    if (currentSong && !currentSong.lyrics && !remoteLyrics.isLoading && !remoteLyrics.isSuccess) {
+      lyrics?.fetchLyrics(currentSong);
+    }
```
This check is not working as expected.
In `lyric-query.ts` -- `useSongLyrics`, the `queryFn` returns null if the server is not Jellyfin.

This causes `remoteLyrics.isSuccess` to always equal true because no error was thrown in the queryFn. The condition will always be falsy and the fetchLyrics is skipped if the song is Navidrome.

Let's split the conditional to make it clearer
```typescript
    const hasTaggedLyrics = currentSong && currentSong.lyrics;
    const hasLyricsResponse =
      !remoteLyrics.isLoading && remoteLyrics?.isSuccess && remoteLyrics?.data !== null;

    if (!hasTaggedLyrics && !hasLyricsResponse) {
      lyrics?.fetchLyrics(currentSong);
    }
```

jeffvli commented on src/main/preload/lyrics.ts:
```diff
@@ -0,0 +1,17 @@
+import { IpcRendererEvent, ipcRenderer } from 'electron';
+import { QueueSong } from '/@/renderer/api/types';
+
+const fetchLyrics = (song: QueueSong) => {
+  ipcRenderer.send('lyric-fetch', song);
+};
+
+const getLyrics = (
+  cb: (event: IpcRendererEvent, songName: string, source: string, lyric: string) => void,
+) => {
+  ipcRenderer.on('lyric-get', cb);
+};
```
These two function names are a bit confusing as to what they actually do.

`fetchLyrics` -- fetches lyrics from remote sources as opposed to from the server. Rename this to `fetchRemoteLyrics` or similar

`getLyrics` -- listens for and retrieves the lyrics response from the remote sources. getLyrics is already used in the api controller, so rename to `retrieveRemoteLyrics` or similar

jeffvli replied:
Thinking about it again, the listener function should probably be `remoteLyricsListener` since `retrieveRemoteLryics` makes it also sound like a fetch function. In the `mpv-player` preload, I had all of the listener functions under a separate named export called `mpvPlayerListener` so it was easy to identify. Since it's combined here, adding `listener` somewhere in the name should be used so it can be easily identified as such.

kgarner7 committed:
address comments

kgarner7 committed:
rename listener function

jeffvli committed:
Disable query if no song present

jeffvli committed:
Update minimum app dimensions

jeffvli committed:
Refactor layout to grid

jeffvli committed:
Adjust lyrics styling / animations

jeffvli committed:
Additional styling changes

jeffvli committed:
Use scale instead of font-size for active lyric

jeffvli committed:
Improve responsive styles for playerbar

jeffvli committed:
Change padding on vertical player

jeffvli committed:
Fix styling on unsync lyrics

jeffvli committed:
Minor fixes on main component
